File: neet_code_all_cpp/math_and_geometry/zigzag_conversion.cpp
#include <iostream>
#include <string>

class SolutionAns {
    public:
        std::string convert(std::string s, int numRows) {
            if(numRows == 1) {
                return s;
            }

            std::string result = "";
            for(int row = 0; row < numRows; ++row) {
                int increment = 2 * (numRows - 1);
                for(int i = row; i < s.length(); i += increment) {
                    result += s[i];
                    // ??
                    if(row > 0 && row < numRows - 1 && i + increment - 2 * row < s.length()) {
                        result += s[i + increment - 2 * row];
                    }
                }
            }

            return result;
        }
};

int main(void) {
    std::pair<std::string,int> case_1 = {"PAYPALISHIRING", 3};
    // => "PAHNAPLSIIGYIR"
    std::pair<std::string,int> case_2 = {"PAYPALISHIRING", 4};
    // => "PINALSIGYAHRPI"
    std::pair<std::string,int> case_3 = {"A", 1};
    // => "A"

    SolutionAns s_ans;

    std::cout << s_ans.convert(case_1.first, case_1.second) << std::endl;
    std::cout << s_ans.convert(case_2.first, case_2.second) << std::endl;
    std::cout << s_ans.convert(case_3.first, case_3.second) << std::endl;
}

File: neet_code_all/src/math_and_geometry/zigzag_conversion.rs
// 解けなかった
struct Solution;
impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        let s = s.chars().collect::<Vec<char>>();
        let n = s.len();
        let num_rows = num_rows as usize;
        let mut grid = vec![vec!['#'; n]; num_rows];

        let mut p = num_rows - 1;
        for r in 0..num_rows {
            let mut i = r;
            while i < n {
                grid[r][i] = s[i];
                i += 2 * p;
            }
            p -= 1;
            if p == 0 {
                p = num_rows - 1;
            }
        }

        let mut result = vec![];
        for r in 0..num_rows {
            for i in 0..n {
                if grid[r][i] != '#' {
                    result.push(grid[r][i]);
                }
            }
        }
        result.iter().collect()
    }
}

// C++の模範解答より
struct SolutionAnsCpp;
impl SolutionAnsCpp {
    // AC
    pub fn convert(s: String, num_rows: i32) -> String {
        if num_rows == 1 {
            return s;
        }
        let num_rows = num_rows as usize;
        let s = s.chars().collect::<Vec<char>>();
        let mut result = "".to_string();
        for row in 0..num_rows as usize {
            let increment = 2 * (num_rows - 1);
            let mut i = row;
            while i < s.len() {
                result.push(s[i]);
                // ??
                if row > 0 && row < num_rows - 1 && i + increment - 2 * row < s.len() {
                    result.push(s[i + increment - 2 * row]);
                }

                i += increment;
            }
        }

        result
    }
}

fn main() {
    let case_1 = ("PAYPALISHIRING".to_string(), 3);
    // => "PAHNAPLSIIGYIR"
    let case_2 = ("PAYPALISHIRING".to_string(), 4);
    // => "PINALSIGYAHRPI"
    let case_3 = ("A".to_string(), 1);
    // => "A"

    /*
    println!("case_1: {}", Solution::convert(case_1.0.clone(), case_1.1));
    println!("case_2: {}", Solution::convert(case_2.0.clone(), case_2.1));
    */
    //println!("case_3: {}", Solution::convert(case_3.0.clone(), case_3.1));

    println!(
        "case_1: {}",
        SolutionAnsCpp::convert(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        SolutionAnsCpp::convert(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        SolutionAnsCpp::convert(case_3.0.clone(), case_3.1)
    );
}
